### ISSUE TITLE
Trying RC6 Cairo

### ISSUE THREAD

**Issue by uwbwsvd** [WaynePhillipsEA/twinbasic#354]
![grafik](https://user-images.githubusercontent.com/41230018/129373023-68bbc488-361f-4edb-8f3e-c7b54cf3cd70.png)
[RC6Cairo.zip](https://github.com/WaynePhillipsEA/twinbasic/files/6983154/RC6Cairo.zip)
 see also #83
tB: v0.10.4273

**Comment by WaynePhillipsEA** [comment]
The debugger is disconnecting at the AddImage call, but this looks to be due to a fault in the RC6 ImageList.AddImage call, causing a double-free of the byte array.

Note that if you build an EXE, it works... but if you attach an x86 debugger to that EXE, you'll see exceptions at these call sites (which you also see in a VB6 built EXE making the same calls).

**Comment by vbRichClient** [comment]
JFYI, this is not a "fault" in the RC6.ImageList.AddImage call.

Instead it seams, that the *direct* passing of the return-value of the wget-function is not handled properly.

E.g. the tB-Debugger has no problem (on my machine, using tB 0.10.4378),
when I explicitely set a "temporary Buffer-Variable" this way:
 ```
   Dim ImgBytes() As Byte
            ImgBytes = wget("http://vbRichClient.com/Downloads/bgSquares.jpg")
        Cairo.ImageList.AddImage "bgForm", ImgBytes
            ImgBytes = wget("http://vbRichClient.com/Downloads/clockFace.png")
        Cairo.ImageList.AddImage "bgClock", ImgBytes
```
The instable behaviour (when no temp-ImgBytes-Buffer is used) might be caused by *not* having defined the STA-model in your Project-Settings (this should IMO be made the default, to avoid cross-thread-marshaling) - because when explicitely set - the Debugger seems to have less problems with the direct passing of the wget-result into the AddImage-method.

As for "internal Dll-Errors" - these are entirely normal and should be of no concern, 
when "gracefully handled" within the external Dll (which I do).

All the tB-Debugger should be "interested in" (when working with COM-Dlls, performing "external MethodCalls"),
is their returned HResult (when differing from S_OK == 0).

Easiest way to be sure, when a certain Debugger-Error or -Crash happens -
is to just run the same code in the VB6-IDE (which has no problems at all, when "stepping through the lines").

modMain.bas:
```
Public fMain As New cfMain

Sub Main()
      Cairo.SetDPIAwareness '<- we use this here, to avoid a manifest-resource
 
      Cairo.ImageList.AddImage "bgForm", wget("http://vbRichClient.com/Downloads/bgSquares.jpg")
      Cairo.ImageList.AddImage "bgClock", wget("http://vbRichClient.com/Downloads/clockFace.png")
  
  fMain.Form.Show 'after loading 2 ImageResources from the web, we now show the MainForm
  
  Cairo.WidgetForms.EnterMessageLoop 'and finally enter the MessagePump (we exit from it, when Cairo.WidgetForms.Count reaches 0)
End Sub

Public Function wget(URL As String) As Byte()
  Dim http As Object: Set http = CreateObject("WinHttp.WinHttpRequest.5.1")
      http.Open "GET", URL, 0: http.Send
  wget = http.ResponseBody
End Function
```
cfMain.cls:
```
Public WithEvents Form As cWidgetForm
Private VLayout As cLayoutVertical, HBand As cLayoutBand
Private WithEvents grpTzOpt As cWidgetBase, imgClock As cwImage

Private Sub Class_Initialize()
  Set Form = Cairo.WidgetForms.Create(, "Hello RC6-Forms", , 1024, 800)
      Form.WidgetRoot.ImageKey = "bgForm"
      Form.WidgetRoot.Zoom = Form.WidgetRoot.CurrentMonitor.Zoom 'ensure DPI-awareness
      Form.IconImageKey = "bgClock"
      Form.SetMinMaxDimensions 800, 700
      Form.TimerInterval = 100 '<- VBA-UserForm-compatible Timer-Prop
      
  Set VLayout = New cLayoutVertical
      VLayout.AttachTo Form
      
  'Add a horizontal band with two cells to the layout
  Set HBand = VLayout.AddBandWithCells(0.3, 0.7) '(CellWidths > 1 are treated as "absolute", values <=1 relative in percent)

  Set grpTzOpt = HBand.Cell(0).BeginGroupVertical("") 'a vertical group-frame (for a few Option-Boxes)
      HBand.Cell(0).AddOptionBox "Local", True '<- pre-select this Option-Button
      HBand.Cell(0).AddOptionBox "Cairo" 'yeah - pun intended... ;-)
      HBand.Cell(0).AddOptionBox "Tokyo"
      HBand.Cell(0).AddOptionBox "Mumbai"
      HBand.Cell(0).AddOptionBox "Honolulu"
  HBand.Cell(0).EndGroup 'tells the VLayout-cell, that a group is complete (BTW, .BeginGroupHorizontal is available too)
  grpTzOpt.Alpha = 0.55 '<- just to demonstrate the "nested Alpha-Rendering-support" of the Widget-Engine
  
  'the second cell in this horizontal band will take up only our ImgBox-Widget
  Set imgClock = HBand.Cell(1).AddImageBox("", "bgClock", 501, 501)  '<- define BG-ImageKey and Canvas-Size
End Sub

Private Sub grpTzOpt_BubblingEvent(Sender As Object, EventName As String, P1, P2, P3, P4, P5, P6, P7) 'OptBtn-Group-Handler
  If EventName = "Click" Then Form.Tag = Sender.Caption 'simply store the current Caption
End Sub

Private Sub Form_Timer()
  Dim UTC As cUTC: Set UTC = New_c.UTC
  Dim T As Date:   T = UTC.SystemUTC
  Select Case Form.Tag '<- we've stored the last clicked OptionBtn-Caption there
    Case "Local":    T = T + UTC.LocalOffs '<- SystemUTC + LocalOffs is equivalent to T = Now
    Case "Cairo":    T = T + UTC.OffsFromStr("+02:00")
    Case "Tokyo":    T = T + UTC.OffsFromStr("+09:00")
    Case "Mumbai":   T = T + UTC.OffsFromStr("+05:30")
    Case "Honolulu": T = T + UTC.OffsFromStr("-10:00")
  End Select
  'note, that the (non-Local)) TimeZones above, are ones without any DayLight-Saving-Offsets
  '(one will have to determine and add those Extra-DST-Offsets for TimeZones where DayLight-Saving applies)
  
  If imgClock.Widget.Tag <> T Then imgClock.Widget.Tag = T Else Exit Sub 'avoid Canvas-refreshs, if there was no change on T

  imgClock.Cls 'clear any Drawings (which were made on top of the existing ClockFace-BG-Image)
    DrawClockHand imgClock.Canvas, 110, 8, &H80, 30 * (Hour(T) Mod 12 + Minute(T) / 60)
    DrawClockHand imgClock.Canvas, 160, 4, &H808000, 6 * (Minute(T) + Second(T) / 60)
    DrawClockHand imgClock.Canvas, 220, 1, &H8000, 6 * (Second(T))
    DrawCenterDot imgClock.Canvas, Cairo.CreateRadialPattern(0, 0, 10, -3.8, -3.8, 0)
  imgClock.Refresh

  Form.Caption = Form.Tag & "-Clock  " & T
End Sub

'**** Cairo-Drawing-Helpers - the CC is the typical Cairo-Context (in C-examples usually named cr)
Private Sub DrawClockHand(CC As cCairoContext, Length, LineWidth, ByVal Color&, AngleDeg)
  CC.SaveAndTransform CC.Surface.Width / 2, CC.Surface.Height / 2, AngleDeg
    CC.DrawLine 0, Length * -0.8, 0, Length * 0.2, False, LineWidth, Color
  CC.Restore
End Sub

Private Sub DrawCenterDot(CC As cCairoContext, Pat As cCairoPattern)
  Pat.AddGaussianStops_TwoColors vbBlack, vbWhite, 1, 0.9
  CC.SaveAndTransform CC.Surface.Width / 2, CC.Surface.Height / 2
    CC.Arc 0, 0, 10
    CC.Fill 0, Pat
  CC.Restore
End Sub
```

HTH

Olaf

**Comment by uwbwsvd** [comment]
I did this:
![grafik](https://user-images.githubusercontent.com/41230018/129485943-44c0e386-2bb9-4300-88b8-93bf71014e32.png)
and it works as expected
Thanks.

"The instable behaviour (when no temp-ImgBytes-Buffer is used) might be caused by not having defined the STA-model in your Project-Settings (this should IMO be made the default, to avoid cross-thread-marshaling) - because when explicitely set - the Debugger seems to have less problems with the direct passing of the wget-result into the AddImage-method."
To change or even set this setting, is this "standard" knowlegde of a VB/tB programmer?

**Comment by Kr00l** [comment]
I vote also my 2 cents that STA is set by default. (Inherited default when no setting is defined)

**Comment by WaynePhillipsEA** [comment]
It absolutely should be STA by default.  Apologies, I hadn't noticed that it wasn't.

This bug actually wasn't due to the threading model, but it was amplified by it due to the way COM marshals arrays.  It was a codegen bug, and a fix is imminent.  Thanks.

**Comment by WaynePhillipsEA** [comment]
Fixed in v0.10.4386, thanks for reporting.   And STA is now the default when not explicitly set.
